### ISSUE TITLE
DistributedDataParallel cannot handle multi-device runs with non-evenly divisible batch sizes

### ISSUE THREAD

**Issue by Flamefire** [pytorch/pytorch#46175]
## 🐛 Bug

When running the test suite after the build the DistributedDataParallelSingleProcessTest fails with `'initialize_buckets' must NOT be called during autograd execution`

Reproducer: `./run_test.py -i distributed/test_c10d_spawn DistributedDataParallelSingleProcessTest.test_rnn`

Edit: Reason is that the batch size is not evenly divisible by the number of devices/GPUs leading to the Reducer expecting more devices participating than there actually are.

Stacktrace:
```
ERROR: test_rnn (__main__.DistributedDataParallelSingleProcessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/easybuild-tmp/eb-qx9yda/tmpnMATEi/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 396, in wrapper
    fn(*args, **kwargs)
  File "distributed/test_c10d_spawn.py", line 308, in test_rnn
    self._test_base(net, inp, check_allclose=False)
  File "distributed/test_c10d_spawn.py", line 248, in _test_base
    ddp_out = ddp(*inp)
  File "/tmp/easybuild-tmp/eb-qx9yda/tmpnMATEi/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/tmp/easybuild-tmp/eb-qx9yda/tmpnMATEi/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 606, in forward
    if self.reducer._rebuild_buckets():
RuntimeError: `initialize_buckets` must NOT be called during autograd execution.
```
 - PyTorch Version (e.g., 1.0): 1.7.0-rc1
 - OS (e.g., Linux): RHEL 7
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Python version: 3.7.4

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski

**Comment by Flamefire** [comment]
I'm also getting related errors in that test:

```
test_cuda (__main__.DistributedDataParallelSingleProcessTest) ... /tmp/easybuild-tmp/eb-GDtJd7/tmpfyxbKC/lib/python3.7/site-packages/torch/nn/parallel/distributed.py:448: UserWarning: Single-Process Multi-GPU is not the recommended mode for DDP. In this mode, each DDP instance operates on multiple devices and creates multiple module replicas within one process. The overhead of scatter/gather and GIL contention in every forward pass can slow down training. Please consider using one DDP instance per device or per module replica by explicitly setting device_ids or CUDA_VISIBLE_DEVICES. 
  "Single-Process Multi-GPU is not the recommended mode for "
ok
test_rnn (__main__.DistributedDataParallelSingleProcessTest) ... /tmp/easybuild-tmp/eb-GDtJd7/tmpfyxbKC/lib/python3.7/site-packages/torch/nn/parallel/_functions.py:64: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
```

Setting `CUDA_VISIBLE_DEVICES=0` before running the test makes it pass. However the other tests in this file require multiple GPUs.

**Comment by Flamefire** [comment]
Ok, after some debugging I found that this is related to the number of GPUs available. With 2 GPUs it passes, with 3 or more it fails. The reason seems to be that at https://github.com/pytorch/pytorch/blob/a814231616f5db7fcdb8793f64c12d652d1e8572/torch/nn/parallel/distributed.py#L571-L572 a `Reducer` instance is created with 1 parameter (called `replica` in C++) per device. This becomes important as at https://github.com/pytorch/pytorch/blob/a814231616f5db7fcdb8793f64c12d652d1e8572/torch/csrc/distributed/c10d/reducer.cpp#L625-L634 the `finalize_backward` method is called, if all replicas finished processing.

However at https://github.com/pytorch/pytorch/blob/a814231616f5db7fcdb8793f64c12d652d1e8572/torch/nn/parallel/distributed.py#L663 the `scatter` distributes the inputs among less than the number of available devices. In this case for 6 available GPUs only 4 are used (found by a `print(str(inputs))` after the scatter and looking at `device=`), reducing to 3 GPUs still uses 2 parts, and for 2 GPUs it will split the inputs on both.

The effect can be seen when adding a print statement on the `replica.pending/bucket.pending` at https://github.com/pytorch/pytorch/blob/a814231616f5db7fcdb8793f64c12d652d1e8572/torch/csrc/distributed/c10d/reducer.cpp#L625 where it will stop decreasing the `bucket.pending` after the actual number of splits are handled (the times `replica.pending` reaches zero)

So to summarize: 
- All GPUs are passed to the `Reducer`
- `scatter` uses only a subset of those GPUs
- `Reducer` expects all to be used, and hence doesn't call the finalize method
- the next `_rebuild_buckets` call fails due to missing finalize call

**Comment by Flamefire** [comment]
And the final link: The `scatter` method splits the input along the given dimension into equal-sized chunks. The problem here: The batch size (split dimension) is `4`. Hence `4 / ceil(4/6) == 1` and `4 / ceil(4/3) == 2` which explains the number of replicas I see.

So the problem is that the `DistributedDataParallel` Module and/or the `Reducer` class can't handle the case that less than the maximum number of devices is used but request equally sized chunks.

**Comment by rohan-varma** [comment]
Thanks for the extensive report and the fix in https://github.com/pytorch/pytorch/pull/46186! As @mrshenli mentioned, we are working towards no longer supporting single-process, multiple device mode in DDP, so I think we can close this issue out given that there is a workaround.
